File: COA software/frontend/src/types/conflict.ts
export type ConflictSeverity = 'INFO' | 'WARNING' | 'HIGH' | 'CRITICAL'

export type ConflictType =
  | 'TRAIN_CONFLICT'
  | 'BLOCK_OVERLAP'
  | 'CORRIDOR_CONFLICT'
  | 'ISOLATION_CONFLICT'
  | 'DEPARTMENT_CONFLICT'
  | 'SAFETY_CONFLICT'
  | 'MAINTENANCE_CONFLICT'
  | 'CAPACITY_CONFLICT'
  | 'OPERATIONAL_BUFFER_CONFLICT'

export interface ConflictItem {
  conflict_type: ConflictType
  severity: ConflictSeverity
  entity_type?: string
  entity_id?: string
  entity_name?: string
  start_time?: string
  end_time?: string
  description: string
  resolution?: string
}

export interface BlockEvaluationData {
  corridor_id: string
  corridor_name?: string
  start_time: string
  end_time: string
  start_datetime: string
  end_datetime: string
  duration_minutes: number
  feasible: boolean
  severity: ConflictSeverity
  conflict_count: number
  critical_conflicts_count: number
  conflicts: ConflictItem[]
  resolution_suggestions: string[]
  affected_trains: any[]
  train_impact: any
  alternatives: any[]
  shared_block_possible: boolean
}

export interface BlockEvaluationPayload {
  corridor_id: string
  start_time: string
  end_time: string
  task_ids?: string[]
  isolation_required?: boolean
  exclude_block_id?: string
}

export interface FeasibleWindowItem {
  start_time: string
  end_time: string
  start_datetime: string
  end_datetime: string
  duration_minutes: number
  impact_score: number
  expected_delay_minutes: number
  conflict_count: number
  severity: ConflictSeverity
  feasible: boolean
  reason: string
}

export interface FeasibleWindowsPayload {
  corridor_id: string
  date: string
  duration_minutes?: number
  preferred_start_hour?: number
  preferred_end_hour?: number
  task_ids?: string[]
}

File: COA software/backend/app/schemas/conflict.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class BlockEvaluationRequest(BaseModel):
    corridor_id: str = Field(..., description="Corridor UUID")
    start_time: datetime = Field(..., description="Block start datetime")
    end_time: datetime = Field(..., description="Block end datetime")
    task_ids: Optional[List[str]] = Field(default=None, description="Optional list of maintenance task UUIDs")
    isolation_required: Optional[bool] = Field(default=False, description="Isolation requirement flag")
    exclude_block_id: Optional[str] = Field(default=None, description="Block request ID to exclude from overlap checks")


class ConflictItem(BaseModel):
    conflict_type: str
    severity: str
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    entity_name: Optional[str] = None
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    description: str
    resolution: Optional[str] = None


class BlockEvaluationData(BaseModel):
    corridor_id: str
    corridor_name: Optional[str] = None
    start_time: str
    end_time: str
    start_datetime: str
    end_datetime: str
    duration_minutes: int
    feasible: bool
    severity: str
    conflict_count: int
    critical_conflicts_count: int
    conflicts: List[ConflictItem] = []
    resolution_suggestions: List[str] = []
    affected_trains: List[Dict[str, Any]] = []
    train_impact: Dict[str, Any] = {}
    alternatives: List[Dict[str, Any]] = []
    shared_block_possible: bool = False


class BlockEvaluationResponse(BaseModel):
    success: bool = True
    data: BlockEvaluationData
    message: str = "Block evaluation completed successfully"


class FeasibleWindowsRequest(BaseModel):
    corridor_id: str
    date: datetime
    duration_minutes: int = 120
    preferred_start_hour: Optional[int] = 0
    preferred_end_hour: Optional[int] = 24
    task_ids: Optional[List[str]] = None


class FeasibleWindowItem(BaseModel):
    start_time: str
    end_time: str
    start_datetime: str
    end_datetime: str
    duration_minutes: int
    impact_score: float
    expected_delay_minutes: float
    conflict_count: int
    severity: str
    feasible: bool
    reason: str


class FeasibleWindowsResponse(BaseModel):
    success: bool = True
    data: List[FeasibleWindowItem] = []
    message: str = "Feasible windows retrieved successfully"
